### PR TITLE
ULTRA Add new task

### PR DESCRIPTION
**Pull request by JenishPatel99**
Goal: Introduce task 3 in ULTRA

ULTRA currently supports task 1 and task 2: Their generalization objectives are listed below
- Task 1: Train on T-intersections and evaluate on Cross-intersection (Generalize across different intersection types)
- Task 2: Train on low/high traffic densities and evaluate on high/low traffic densities (Generalize across different intersection types)

The generalization objective of new task:
- Task 3: Train on roads with n-number of lanes and evaluate on roads with (n+k)-number of lanes ({k ∈ Ζ | 1< k ≤ 3}). (Generalize across different number of lanes)

Task 3 includes 3 different levels (easy, medium, hard). Like task 1, the levels are constructed to enforce a difficulty range (i,e, each level will have its own "difficulty" setting).

| Level | Training set (10000 scenarios) | Testing set (200 scenarios) |  
| ------------- |:-------------:| :-----:|
| easy      | 2 lane maps | 3 lane maps (k=1) |
| medium      | 2 lane maps | 4 lane maps (k=2) |
| hard  | 2 lane maps | 5 lane maps (k=3) |

The parameters that are variable in the training and testing set
- Intersection shape (T or cross)
- Speed limit

Unlike, task 1, where the traffic distribution was different across the levels, task 3 will have a uniform traffic distribution.

| Traffic distribution | percent (%) |
| ---- | :---: |
| no-traffic | 61 |
| low-density | 33 |
| mid-density | 3 |
| high-density | 6 |

### CODE REVIEW

File: ultra/ultra/config.yaml [hunk 1]
```diff
@@ -59,3 +59,13 @@ tasks:
...
+  task3:
+    easy:
+      train: "scenarios/task3/train_easy*"
+      test:  "scenarios/task3/test_easy*"
+    medium:
+      train: "scenarios/task3/train_medium*"
+      test:  "scenarios/task3/test_medium*"
+    hard:
+      train: "scenarios/task3/train_hard*"
+      test:  "scenarios/task3/test_hard*"
```
Comment (christianjans) [inline ultra/ultra/config.yaml:71]:
```suggestion
      test:  "scenarios/task3/test_hard*"

```

File: ultra/ultra/scenarios/task3/config.yaml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+           stops: null
+           bubbles: null
+        5lane_c:
+           percent: 0.5
+           specs: [[50kmh,low-density,0.21],[70kmh,low-density,0.20],[100kmh,low-density,0.20], #61%
+                   [50kmh,mid-density,0.11],[70kmh,mid-density,0.11],[100kmh,mid-density,0.11], #33%,
+                   [50kmh,no-traffic,0.01],[70kmh,no-traffic,0.01],[100kmh,no-traffic,0.01], #3%
+                   [50kmh,high-density,0.01],[70kmh,high-density,0.01],[100kmh,high-density,0.01]] # 3%
+           stops: null
+           bubbles: null
```
Comment (christianjans) [inline ultra/ultra/scenarios/task3/config.yaml:145]:
```suggestion
           bubbles: null

```

**Review by christianjans — COMMENTED** [review]
I think this looks fine, I am just wondering if we should perform some analysis on these new scenarios before we merge, or have we already done this? Do you think this is needed?

Also, were you still going to implement the "side task" mentioned in this PR's description? I know we had talked about the feasibility of this test earlier, but have you come up with a solution?

**Comment by JenishPatel99** [comment]
> I think this looks fine, I am just wondering if we should perform some analysis on these new scenarios before we merge, or have we already done this? Do you think this is needed?
> 
> Also, were you still going to implement the "side task" mentioned in this PR's description? I know we had talked about the feasibility of this test earlier, but have you come up with a solution?

I haven't yet ran any experiments with our baselines. I will be not merging this branch until we are done testing task3. In terms of analysis, I believe that all of the scenarios generated should be highly passable (i.e. for 10000 scenarios per level, the success rate of scenarios is above 90%). 

I have not found a suitable method to test all of tasks and levels without changing the configs. It is not a high priority as of right now, but I will keep thinking about it

**Review by christianjans — APPROVED** [review]

**Pull request merged**